examples/training/resnet50/main.cpp

```diff
@@ -0,0 +1,253 @@
+                     clEnumValN(BackendKind::OpenCL, "opencl", "Use OpenCL")),
+    llvm::cl::init(BackendKind::Interpreter), llvm::cl::cat(category));
+
+/// Finds the maximum value and its index within \p N range.
+size_t findMaxIndex(const Tensor &t, size_t range) {
+  auto handle = t.getHandle<>();
+  size_t ret{0};
+  float maxVal{handle.raw(0)};
+  for (size_t c = 0; c < 10; ++c) {
+    auto val = handle.raw(c);
+    if (maxVal <= val) {
+      maxVal = val;
+      ret = c;
+    }
+  }
+  return ret;
+}
+} // namespace
+
+int main(int argc, char **argv) {
+  llvm::cl::ParseCommandLineOptions(argc, argv,
+                                    " ResNet50 Training Example\n\n");
+
+  Module Mod;
+
+  // Load ResNet model.
+  Function *F = Mod.createFunction("resnet50");
+
+  llvm::outs() << "Loading resnet50 model.\n";
+
+  // We expect the input to be NCHW.
+  llvm::ArrayRef<size_t> AllImagesDims = {CIFAR_NUM_IMAGES, IMAGE_COLORS,
+                                          IMAGE_HEIGHT, IMAGE_WIDTH};
+  llvm::ArrayRef<size_t> BatchImagesDims = {miniBatchSize, IMAGE_COLORS,
+                                            IMAGE_HEIGHT, IMAGE_WIDTH};
+  llvm::ArrayRef<size_t> InitShapeDims = {1, IMAGE_COLORS, IMAGE_HEIGHT,
+                                          IMAGE_WIDTH};
+
+  llvm::ArrayRef<size_t> AllLabelsDims = {CIFAR_NUM_IMAGES, LABELS_SIZE};
+
+  PlaceholderBindings Bindings;
+  const char inputName[] = "gpu_0/data";
+  TypeRef inputType = Mod.uniqueType(ElemKind::FloatTy, InitShapeDims);
+  llvm::Error errPtr = llvm::Error::success();
+  Caffe2ModelLoader loader("resnet50/predict_net.pbtxt", "resnet50/init_net.pb",
+                           {inputName}, {inputType}, *F, &errPtr, &Bindings);
+
+  // Loader has been taken care of the weights randomization.
+
+  // Training
+  TrainingConfig TC;
+  BackendKind ExecutionBackend{executionBackend};
+  ExecutionEngine EE(ExecutionBackend);
+
+  TC.learningRate = learningRate;
+  TC.momentum = momentum;
+  TC.L2Decay = l2Decay;
+  TC.batchSize = miniBatchSize;
+
+  llvm::outs() << "Creating input/output placeholders.\n";
+
+  // Create the input layer.
+  auto *A =
+      Mod.createPlaceholder(ElemKind::FloatTy, BatchImagesDims, "input", false);
+  Bindings.allocate(A);
+
+  // Create the output layer.
+  auto *E = Mod.createPlaceholder(
+      ElemKind::FloatTy, {miniBatchSize, LABELS_SIZE}, "expected", false);
+  Bindings.allocate(E);
+
+  llvm::outs() << "Differentiating graph.\n";
+  Function *TF = glow::differentiate(F, TC);
+  EE.compile(CompilationMode::Train, TF);
+
+  // Allocate placeholders without tensors.
+  Bindings.allocate(Mod.getPlaceholders());
+
+  // Get output tensor.
+  auto *ResultPH = Mod.getPlaceholderByName("save_gpu_0_softmax");
+  auto *Result = Bindings.get(ResultPH);
+
+  llvm::outs() << "Loading the CIFAR-10 database.\n";
+  /// ResNet expects ImageChannelOrder::BGR, ImageLayout::NCHW
+  std::ifstream dbInput("cifar-10-batches-bin/data_batch_1.bin",
```

Comment:
There are plenty assumptions in the code about CIFAR images format and number of images in the batch, if caller will provide an arbitrary file it can easily breaks the internal logic.